wai-extra/Network/Wai/Middleware/RealIp.hs

```diff
@@ -0,0 +1,79 @@
+-- | Infer the remote IP address using headers
+module Network.Wai.Middleware.RealIp
+    ( realIp
+    , realIpHeader
+    , realIpTrusted
+    , defaultTrusted
+    ) where
+
+import qualified Data.ByteString.Char8 as B8 (unpack, split)
+import qualified Data.IP as IP
+import Data.Maybe (fromMaybe, mapMaybe, listToMaybe)
+import Network.HTTP.Types (HeaderName, RequestHeaders)
+import Network.Wai
+import Text.Read (readMaybe)
+
+-- | Infer the remote IP address from the @X-Forwarded-For@ header,
+-- trusting requests from any private IP address. See 'realIpHeader' and
+-- 'realIpTrusted' for more information and options.
+--
+-- @since 3.1.5
+realIp :: Middleware
+realIp = realIpHeader "X-Forwarded-For"
+
+-- | Infer the remote IP address using the given header, trusting
+-- requests from any private IP address. See 'realIpTrusted' for more
+-- information and options.
+--
+-- @since 3.1.5
+realIpHeader :: HeaderName -> Middleware
+realIpHeader header = realIpTrusted header defaultTrusted
+
+-- | Infer the remote IP address using the given header, but only if the
+-- request came from an IP in one of the trusted ranges. The last
+-- non-trusted address is used to replace the 'remoteHost' in the
+-- 'Request', unless all present IP addresses are trusted, in which case
+-- the first address is used. Invalid IP addresses are ignored, and the
+-- remoteHost value remains unaltered if no valid IP addresses are
+-- found.
+--
+-- @since 3.1.5
+realIpTrusted :: HeaderName -> [IP.IPRange] -> Middleware
```

Comment:
Just a thought: would it be more flexible to provide a predicate argument here to check the IP address, with the current `ipInRange` as a helper function in this module?

Reply:
That's a good idea. So we would have
```
realIpTrusted :: HeaderName -> (IP.IP -> Bool) -> Middleware
```
with an example implementation being
```
realIpTrusted "X-Forwarded-For" (\ip -> any (ipInRange ip) defaultTrusted)
```

(I'll see if I can rework things to avoid the lambda.)